---
title: 'Data Analysis: Investigating the population genetics of Mytilus californianus
  using a coalescent sampler'
author: "Vanessa Garcia"
date: "5/06/2021"
output:
  html_document:
    keep_md: yes
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 
```{r}
install.packages("knitr")
install.packages("pegas")
install.packages("rentrez")
install.packages("stringr")
install.packages("plyr")
install.packages("strataG")
```

```{r}
library(knitr)
library(ape)
library(pegas)
library(rentrez)
library(stringr)
library(plyr)
library(muscle)
#library(strataG)
library(magrittr)
library(dplyr)
library(Biostrings)
library(BiocGenerics)
```





```{r}
#dog.aln<-read.dna("mcali_revised.fasta", format="fasta", as.matrix=T)
myt_aln<-read.FASTA("lat_sorted_seqs.fasta")
myt_aln<-read.FASTA("mcalifornianus_mtDNA_alphapops_210514.fasta")
myt_aln

```

#Seg Sites
```{r}
sites <-seg.sites(myt_aln)
sites
```


```{r}
nss <-length(sites)
nss
```


```{r}
myt_aln<-as.matrix(myt_aln) #do I need to turn the list into a matrix? EDC-YES

# Pop Extract Code
pop<-gsub(labels(myt_aln),pattern="([A-Za-z_]+)_\\d+",replacement="\\1")
pop<-gsub(pop,pattern="_",replacement="")
#rownames(myt_aln)<-pop
```

```{r}
#visualize variation in alignment
image.DNAbin(myt_aln,cex.lab = .2)
```



```{r}
#Haplotype Code
d<-dist.dna(myt_aln) # I think this is the right matrix to use
h<-pegas::haplotype(myt_aln)
h<-sort(h,what="label")
net<-pegas::haploNet(h)
i<-utils::stack(setNames(attr(h,"index"),rownames(h)))
i<-i[order(i$values),]
ind.hap<-table(hap=i$ind, pop=pop)

ghost.stories<- c("darkslategrey","darkslateblue", "blue4", "royalblue4", "deepskyblue", "royalblue1", "skyblue4","steelblue", "lightsteelblue1", "lightskyblue2", "mintcream","lightgoldenrod","lightgoldenrod2")

plot(net,size=attr(net,"freq"),scale.ratio=2.25,pie=ind.hap,legend=F,labels=F,threshold=0,show.mutation=2,bg=ghost.stories,pch=19,ncol=1,cex=0.85)
legend("topleft",colnames(ind.hap),col=ghost.stories,pch=19,ncol=1,cex=0.85)
replot()
```


```{r}
h
```



```{r}
#Haplotype Network- how many subs would be required to go from one to another hap

#install.packages("rmarkdown")
#library(rmarkdown)
net<-haploNet(h)
fq<-attr(net,"freq")
plot(net,size=fq, threshold=0, cex=.8)

sample.no<-attr(h, "index")
pop.by.haplo<-sapply(sample.no, function(x) rownames(myt_aln)[x])
sorted<-sapply(pop.by.haplo,paste, collapse="")
haplo<-as.roman(1:153)
queue<-data.frame(haplotype=as.character(haplo), Population=sorted)
queue
myt.div<-nuc.div(myt_aln)
```


<<<<<<< HEAD
```{r}
#install.packages("pegas")
library(pegas)
#this doesn't stratify by population
myt.div<-nuc.div(myt_aln)
```
=======
>>>>>>> 9cf3374a40fdd366590a4cd04ad3e5fe3d6f5f85


#Diversity Statistics
```{r}

stratastat<-function(x,pop=pop,fun=nuc.div){
  #this function will calculate stats for a DNAbin object
  #(x), stratified across populations given in a pop.
  #Some functions this will work with: nuc.div()
  #theta.s(), tajima.test() from pegas, FusFs(), exptdHet() from StrataG
stats<-NULL
for(p in unique(pop)){
  stat<-fun(x[grep(p,rownames(x)),])
  stats<-c(stats,stat)
}
names(stats)<-unique(pop)
return(stats)
}

#Haplotype Diversity
myt_aln_b<- myt_aln

hapdivs<-stratastat(myt_aln_b, pop=pop, fun=hap.div)
hapdivs

#Nucleotide Diversity
nucdivs<-stratastat(myt_aln_b, pop=pop, fun=nuc.div)
nucdivs

#Check pegas work with StrataG
which(duplicated(rownames(myt_aln_b)))
rownames(myt_aln_b)[207]

myt_g<-sequence2gtypes(myt_aln_b,strata=pop)
myt_g<-labelHaplotypes(myt_g)
myt_g

samples<-summary(myt_g$gtypes)$strata.smry[,1]
haplotypes<-summary(myt_g$gtypes)$strata.smry[,3]

#make a table of summary stats
diversities<-cbind("# Samples"=samples, "# Haplotypes"=haplotypes, "Nucleotide Diversity"=nucdivs, "Haplotype"=hapdivs)
kable(diversities, digits=3)

```


```{r}
#install.packages("knitr")
#library(knitr)
#PairwisePhi
pairwise_phi<-pairwiseTest(myt_g$gtypes,stats="phist",nrep=1000,quietly = T,model="raw")
phi.table<-kable(pairwise_phi$pair.mat$PHIst, digits = 3)
phi.table
#PHIST
phiST<- pairwise_phi$pair.mat$PHIst
phiST[upper.tri(phiST)] = t(phiST)[upper.tri(phiST)]
heatmap(phiST)
```






#AMOVA
```{r}
# make copy of data to practice sorting
#amova_presorted_seqs<-read.FASTA("mcali_seqs_210506.txt")
myt_amova<-read.FASTA("lat_sorted_seqs.fasta")
myt_amova<-as.matrix(myt_amova)

```


```{r}

#myt_amova<-read.FASTA("labeled_for_hyp_tests.fasta")
#myt_amova<-as.matrix(myt_amova) #do I need to turn the list into a matrix? EDC-YES

# Pop Extract Code
#pop<-gsub(labels(myt_amova),pattern="([A-Za-z_]+)_\\d+",replacement="\\1")
#pop<-gsub(pop,pattern="_",replacement="")


write.csv(pop,"amova_hypotheses.csv",row.names = F)
#Open this file up in a spreadsheet program and fill out your hypotheses for higher level hierarchies, with the name of your hypothesis in the header, and pop as the name of the population column


```


```{r}
#read them back into R
amovahyps<-read.csv("amova_hypotheses1.csv")
pt_conception<-amovahyps$pt_conception
cape_mendo<-amovahyps$cape_mendo
#marko<-amovahyps$marko_us this is basically same as cape mendo except without criego et al. samples
pop1<-as.factor(pop)
pop1

#Calculate the p-distance among all sequences
dists<-dist.dna(myt_amova,model="raw")
#dists<-dist.dna(myt_amova) # I think this is the right matrix to use

```


#Test Pt. Conception Hypothesis
```{r}

amova_out<-pegas::amova(formula=dists~pt_conception/pop1,nperm=1)

amova_out<-pegas::amova(formula=dists~ptconception/pop1,nperm=1000)



FCT<-amova_out$varcomp[1,1]/sum(amova_out$varcomp[,1])
FCTp<-amova_out$varcomp[1,2]
     
FSC<-amova_out$varcomp[2,1]/(amova_out$varcomp[2,1]+amova_out$varcomp[3,1])
FSCp<-amova_out$varcomp[2,2]
      
FST<-(amova_out$varcomp[1,1]+amova_out$varcomp[2,1])/sum(amova_out$varcomp[,1])
FSTp<-NA

pt_conception_result<-c(FCT,FSC,FST)
```

#Test Cape Mendocino Hypothesis
```{r}

amova_out<-pegas::amova(formula=dists~cape_mendo/pop1,nperm=1)

amova_out<-pegas::amova(formula=dists~capemendo/pop1,nperm=1000)



FCT<-amova_out$varcomp[1,1]/sum(amova_out$varcomp[,1])
FCTp<-amova_out$varcomp[1,2]
     
FSC<-amova_out$varcomp[2,1]/(amova_out$varcomp[2,1]+amova_out$varcomp[3,1])
FSCp<-amova_out$varcomp[2,2]
      
FST<-(amova_out$varcomp[1,1]+amova_out$varcomp[2,1])/sum(amova_out$varcomp[,1])
FSTp<-NA

cape_mendo_result<-c(FCT,FSC,FST)

```




```{r}
AMOVA_result<-cbind(c("Among Regions","Among Populations within Regions", "Among Populations"),pt_conception_result,cape_mendo_result)

kable(AMOVA_result)
```


#Isolation by Distance
```{r}
#write it out and read it back in so you don't need to calculate it again
#write.csv(phiST,"phi_st.csv")
phiST2 <- phiST #read.csv("phi_st.csv")
phi.dist <- as.dist(phiST2)

myt_metadata<-read.csv("myt_metadata.csv")
myt_unique<-unique(myt_metadata[,c("locality","decimalLatitude","decimalLongitude")]) #pull out just locality and lat/longs, and get unique values
myt_unique<-myt_unique[order(myt_unique$locality),] #put them in alphabetical order

myt_unique
```

```{r}
install.packages("raster")
library(raster)
gdist<-pointDistance(p1=myt_unique[,c(3,2)], lonlat=T,allpairs=T) #measure great circle distance between points in meters
gdist<-gdist/1000 # convert to km
gdist<-as.dist(gdist) #convert to distance object
```
```{r}
#pull out PhiST matrix and Mantel's test
#pair_phi<-as.dist(pairwise_phi$pair.mat$PHIst) #convert phi_st to distance object
#pair_phi<-pair_phi/(1-pair_phi) #linearize PhiST
install.packages("ade4")
library(ade4)
phi_mantel<-mantel.randtest(pair_phi,gdist)

phi_mantel
```

```{r}
plot(gdist,pair_phi)
identify(gdist,pair_phi) # use this if you want to identify outliers

install.packages("ggplot")
library(ggplot2)
phi_plot<-ggplot(as.data.frame(cbind(gdist,pair_phi)), aes(x=gdist,y=pair_phi)) + geom_point() + geom_smooth(method=lm) + xlab("Great Circle Distance(km)") + ylab(expression(Phi["ST"]/1-Phi["ST"]))
phi_plot
```


#Migrate
```{r}
install.packages(c("phyclust", "phangorn"))
install.packages("TeachingPopGen")
library(phyclust)
library(phangorn)
library(knitr)

tf_co1<-read.dna("consensus_sequences_TEST.fasta", format="fasta",as.matrix=TRUE)
tf_co1
co1_phy<-phyDat(tf_co1)
co1_model1test<-co1_modeltest[order(co1_model1test$BIC),]
head(co1_model1test)

env<-attr(co1_model1test,"env")
HKY<-get("HKY+G",env)
eval(HKY,env=env) #shape param is .3338988)
plot(density(rgamma(1000,shape=0.2194787),xlim=c(0,1)))

phangorn:::discrete.gamma(0.3338988,4)

TiTvRatio((tf_co1),"./mytilus_migrate_2.nex",format="DNA",interleaved=F,missing="N")

```

#Migrate-n
```{r}
#setwd()
setwd("~/Desktop/migrate3")
marglike<-
  data.frame(model=character(0), thermodynamic=numeric(0),bezier.corrected=numeric(0),harmonic=numeric(0),stringsAsFactors = F)#initialize a df to take the values
l=1
for(i in list.files()){
  wd3<-file.path("~/Desktop/migrate3",i)
  print(wd3)
  if(!file.exists((wd3))){next}
  setwd(wd3)
  outfile<-scan(file="outfile.txt",what="character",sep="\n")
  
  thermoline<-grep("(1a)",outfile,value=T)
  if(length(thermoline)==0){next}
  thermoline<-strsplit(thermoline,split="=",fixed=T)
  thermo<-as.numeric(substr(thermoline[[1]][2],start=1,stop=12))
  bezistart<-grep("\\(",strsplit(thermoline[[1]][2],split="")[[1]])+1
  bezier<-as.numeric(substr(thermoline[[1]][2],start=bezistart,stop=bezistart+11))
  
  harmoline<-grep("\\(2\\) H",outfile,value=T)
  harmo<-as.numeric(substr(harmoline[[1]][2],start=1,stop=12))
  marglike[l,]<-c(i,thermo,bezier,harmo)
  
  }
```

##calculating log bayes factors
```{r}
bfcalcs<-function(df,ml="bezier.corrected"){
  df$thermodynamic<-as.numeric((df$thermodynamic))
  df$bezier.corrected<-as.numeric(df$bezier.corrected)
  df$harmonic<-as.numeric(df$harmonic)
  mlcol<-df[,ml]
  bmvalue<-mlcol[which.max(mlcol)]
  lbf<-2*(mlcol-bmvalue)
  choice<-rank(-mlcol)
  modelprob<-exp(lbf/2)/sum(exp(lbf/2))
  dfall<-cbind(df,lbf,choice,modelprob)
  return(dfall)
}
```
#model selection
```{r}
bftable<-bfcalcs(marlike)
kable(bftable)
write.csv(bftable,"model_selection_table.csv")
```

