---
title: 'Data Analysis: Investigating the population genetics of Mytilus californianus
  using a coalescent sampler'
author: "Vanessa Garcia"
date: "2/19/2021"
output:
  html_document:
    keep_md: yes
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 
```{r}
install.packages("knitr")
install.packages("pegas")
install.packages("rentrez")
install.packages("stringr")
install.packages("plyr")
#install.packages("strataG")
```

```{r}
library(knitr)
library(ape)
library(pegas)
library(rentrez)
library(stringr)
library(plyr)
library(muscle)
install.packages("strataG")
library(strataG)
library(magrittr)
library(dplyr)
library(Biostrings)
library(BiocGenerics)
```

```{r}
#mussel.dat<-read.FASTA("mcali_alignment_030521.fasta")
```
#The samples must be aligned and trimmed so that they are all the same length. 

##I did this in Geneious but I couldn't figure out how to transfer the annotations from the multiple alignment back to the consensus sequences of the individual samples before exporting the Fasta doc... work on this more
```{r}
#class(mussel.dat)
#convert DNAbin to DNAStringSet
#install.packages("magrittr")
#library(magrittr)


#converts from dnabin to dnastringset for muscle function
#mussel.string= ((mussel.dat %>% as.character %>% lapply(.,paste0,collapse="") %>% unlist %>% DNAStringSet))


#align to make same length
mussel.aln<-muscle(mussel.string)

```

```{r}
#visualize variation in alignment
image.DNAbin(mussel.dat,cex.lab = .2)
```

##test alignment with read.dna -> so this imports the already aligned data and puts it into a matrix
```{r}
dog.aln<-read.dna("mcali_alignment_030521.fasta", format="fasta", as.matrix=T)

```


#Measuring Genetic Diversity
```{r}
#Number of Segregating Sites
mussel.sites<-seg.sites(dog.aln)# don't need to sub bc I manually edited fasta so that seqs were same length
mussel.ss<-length(mussel.sites)
mussel.sites
mussel.ss
#output obviously shows that data is misaligned and we can also see that from the viz above.Need to import correct data in correct format/alignment
#we want to focus on the mussell.ss sites that show some degree of variation
```
```{r}
#Average pairwise differences
seqs<-sample(1:289,2) #change x to reflect sample size; this collects 2 random samples from sample set
seqs

#segregating sites
ss<-(seg.sites(dog.aln[seqs,]))
ss

ss_tot<-length(seg.sites(dog.aln[seqs,]))
ss_tot

mussel.ndiv<-nuc.div(dog.aln)
mussel.ndiv #average diversity per nucleotide. If we pick a rando position from two rando seqs, this is the probability that they would be diff bases

mussel.pi<-ncol(dog.aln)*mussel.ndiv
mussel.pi #avg num of diff bw any two seqs
```

haplotype: haploid genotype; string of bases that are found on one chromosome since we are only looking at one locus
```{r}
#Haplotype Diversity
mussel.hap<-haplotype(dog.aln)
mussel.hap

mussel.hd<-hap.div(dog.aln)
mussel.hd # prob that two rando seqs in sample will be diff

```

```{r}
#Visualizing Seq Variation -- created tree in Geneious
#mussel.dist<-dist.dna(dog.aln) #calc dist matrix

#plot(nj(mussel.dist), type="phylo",cex=.8) #plot a neighbor joining tree
#add.scale.bar(x=.005, y=1, length =.0016)
```
```{r}
#Folded Site Frequency Spectrum
mytilus_spectrum<- site.spectrum(mussel.dat)
plot(mytilus_spectrum,col="navy",main= "Folded Site Frequency Spectrum",
     xlab="Frequency", ylab="Number of SNPs")
```


```{r}
# Pop Extract Code
pop<-gsub(labels(dog.aln),pattern="([A-Za-z_]+)_\\d+",replacement="\\1")
pop<-gsub(pop,pattern="_",replacement="")
length(pop)
```


```{r}
#Haplotype Code
d<-dist.dna(dog.aln) # I think this is the right matrix to use
h<-pegas::haplotype(dog.aln)
h<-sort(h,what="label")
net<-pegas::haploNet(h)
i<-utils::stack(setNames(attr(h,"index"),rownames(h)))
i<-i[order(i$values),]
ind.hap<-table(hap=i$ind, pop=pop)

ghost.stories<- c("darkslategrey","darkslateblue", "blue4", "royalblue4", "deepskyblue", "royalblue1", "skyblue4","steelblue", "lightsteelblue1", "lightskyblue2", "mintcream","lightgoldenrod","lightgoldenrod2")

  #c("lightsteelblue1","royalblue4", "royalblue1","steelblue","deepskyblue", "lightskyblue2", "ghostwhite","skyblue4","mintcream","lightgoldenrod","darkslategrey","darkslateblue","blue4")

plot(net,size=attr(net,"freq"),scale.ratio=2.25,pie=ind.hap,legend=F,labels=F,threshold=0,show.mutation=2,bg=ghost.stories,pch=19,ncol=1,cex=0.85)
legend("topleft",colnames(ind.hap),col=ghost.stories,pch=19,ncol=1,cex=0.85)
replot()
```





```{r}
#Haplotype Network- how many subs would be required to go from one to another hap

#install.packages("rmarkdown")
#library(rmarkdown)
net<-haploNet(mussel.hap)
fq<-attr(net,"freq")
plot(net,size=fq, threshold=0, cex=.8)

sample.no<-attr(mussel.hap, "index")
pop.by.haplo<-sapply(sample.no, function(x) rownames(dog.aln)[x])
sorted<-sapply(pop.by.haplo,paste, collapse="")
haplo<-as.roman(1:152)
kable(data.frame(haplotype=as.character(haplo), Population=sorted))
```


#Creating Haplotype Network in Popart
```{r}
#Prep for Popart; this needs to be adjusted for my own data
hapseqs<-h[h] # pull out the DNAbin object from the haplotype object and write into nexus format
write.nexus.data(hapseqs,
                 file="mcali_haplotypes.nex",
                 interleaved=F)
#write both haplotypes to csv
write.csv(ind.hap,
          file="",
          quote=F)

#write the flattened metadata too
write.csv(event_info,
          file="",
          quote=F)
```

#Charles div stats
```{r}
#co1.mat<-read.FASTA("mcali_alignment_030521.fasta")
co1.2<-dog.aln
co1.2g<-sequence2gtypes(co1.2,strata=pop)
co1.2g
```

```{r}
co1.g<-labelHaplotypes(co1.2g)
co1.g

pair_phi<-pairwiseTest(co1.2g,stats="phist",nrep=1000,quietly = T,model="raw")
phi.table<-kable(head(pair_phi$pair.mat$PHIst))
phi.table
phi.matrix<-as.matrix(phi.table)
write.csv(phi.matrix,"phi_st.csv")
```

```{r}
#F-stats
pairwise_F<-pairwiseTest(co1.2g,stats="phist",nrep=1000,quietly = T)
kable(head(pairwise_F$pair.mat$Fst))

statPhist(co1.2g)
statFst(co1.2g)
```

#Diversity Statistics
```{r}
#create copy of dataset 
co1.3<-co1.2
rownames(co1.3)<-pop
stratastat<-function(x,pop=pop,fun=nuc.div){
  #this function will calculate stats for a DNAbin object
  #(x), stratified across populations given in a pop.
  #Some functions this will work with: nuc.div()
  #theta.s(), tajima.test() from pegas, FusFs(), exptdHet() from StrataG
stats<-NULL
for(p in unique(pop)){
  stat<-fun(x[grep(p,rownames(x)),])
  stats<-c(stats,stat)
}
names(stats)<-unique(pop)
return(stats)
}

#try it out
#hapdivs<-stratastat(data_b, pop=pop, fun=hap.div)
#hapdivs

#Nucleotide Diversity
nucdivs<-stratastat(co1.3, pop=pop, fun=nuc.div)
nucdivs

#Check pegas work with StrataG

data_g<-sequence2gtypes(co1.mat,strata=pop)
#data_g<-labelHaplotypes(co1.2g)
data_g

samples<-summary(co1.2g$gtypes)$strata.smry[,1]
haplotypes<-summary(co1.2g$gtypes)$strata.smry[,3]

#make a table of summary stats
diversities<-cbind("# Samples"=samples, "# Haplotypes"=haplotypes, "Nucleotide Diversity"=nucdivs, "Haplotype"=hapdivs)
kable(diversities, digits=3)

```
#F-Statistics
```{r}
#this stat gives us the proportion of nucleotide diversity among subpopulations, relative to the total
#Pairwise Stats
pairwise_phi<-pairwiseTest(data_g$gtypes,stats="phist",nrep=1000,quietly=T,model="raw")

#PhiST in lower triangle, p-value in upper triangle
kable(pairwise_phi$pair.mat$PHIst, digits=3)
phiST<-pairwise_phi$pair.mat$PHIst
phiST[upper.tri(phiST)]= t(phiST)[upper.tri(phiST)]
heatmap(phiST)
```

#Migrate
```{r}
install.packages(c("phyclust", "phangorn"))
install.packages("TeachingPopGen")
library(phyclust)
library(phangorn)
library(knitr)

tf_co1<-read.dna("consensus_sequences_TEST.fasta", format="fasta",as.matrix=TRUE)
tf_co1
co1_phy<-phyDat(tf_co1)
co1_model1test<-co1_modeltest[order(co1_model1test$BIC),]
head(co1_model1test)

env<-attr(co1_model1test,"env")
HKY<-get("HKY+G",env)
eval(HKY,env=env) #shape param is .3338988)
plot(density(rgamma(1000,shape=0.2194787),xlim=c(0,1)))

phangorn:::discrete.gamma(0.3338988,4)

TiTvRatio((tf_co1),"./mytilus_migrate_2.nex",format="DNA",interleaved=F,missing="N")

```

#Migrate-n
```{r}
#setwd()
setwd("~/Desktop/migrate3")
marglike<-
  data.frame(model=character(0), thermodynamic=numeric(0),bezier.corrected=numeric(0),harmonic=numeric(0),stringsAsFactors = F)#initialize a df to take the values
l=1
for(i in list.files()){
  wd3<-file.path("~/Desktop/migrate3",i)
  print(wd3)
  if(!file.exists((wd3))){next}
  setwd(wd3)
  outfile<-scan(file="outfile.txt",what="character",sep="\n")
  
  thermoline<-grep("(1a)",outfile,value=T)
  if(length(thermoline)==0){next}
  thermoline<-strsplit(thermoline,split="=",fixed=T)
  thermo<-as.numeric(substr(thermoline[[1]][2],start=1,stop=12))
  bezistart<-grep("\\(",strsplit(thermoline[[1]][2],split="")[[1]])+1
  bezier<-as.numeric(substr(thermoline[[1]][2],start=bezistart,stop=bezistart+11))
  
  harmoline<-grep("\\(2\\) H",outfile,value=T)
  harmo<-as.numeric(substr(harmoline[[1]][2],start=1,stop=12))
  marglike[l,]<-c(i,thermo,bezier,harmo)
  
  }
```

##calculating log bayes factors
```{r}
bfcalcs<-function(df,ml="bezier.corrected"){
  df$thermodynamic<-as.numeric((df$thermodynamic))
  df$bezier.corrected<-as.numeric(df$bezier.corrected)
  df$harmonic<-as.numeric(df$harmonic)
  mlcol<-df[,ml]
  bmvalue<-mlcol[which.max(mlcol)]
  lbf<-2*(mlcol-bmvalue)
  choice<-rank(-mlcol)
  modelprob<-exp(lbf/2)/sum(exp(lbf/2))
  dfall<-cbind(df,lbf,choice,modelprob)
  return(dfall)
}
```
#model selection
```{r}
bftable<-bfcalcs(marlike)
kable(bftable)
write.csv(bftable,"model_selection_table.csv")
```

